gpio: microchip: improve driver DTS configurability

fabled opened the pull request:
This fixes the driver to be fully configured from DTS, and puts
the GIRQ and valid pin configuration to DTS also. This makes the
driver more suitable for other SoCs in the same family.

Also read and write functions are fixed to not use peripheral
base address from HAL, but from DTS.

This is in preparation for supporting CEC1702.

@scottwcpg @franciscomunoz @galak

fabled committed:
gpio: mchp xec: make fully DTS configurable

This fixes the driver to be fully configured from DTS, and puts
the GIRQ and valid pin configuration to DTS also. This makes the
driver more suitable for other SoCs in the same family.

Also read and write functions are fixed to not use peripheral
base address from HAL, but from DTS.

Signed-off-by: Timo Teräs <timo.teras@iki.fi>

galak commented on dts/bindings/gpio/microchip,xec-gpio.yaml:
```diff
@@ -23,6 +23,16 @@ properties:
     "#gpio-cells":
       const: 2
 
+    valid-bitmap:
```
We should use `gpio-reserved-ranges` property to convey this information.

galak reviewed (changes requested):
We need to look at supporting `gpio-reserved-ranges`.  See https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt#L150

fabled commented:
> We need to look at supporting `gpio-reserved-ranges`. See https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt#L150

Thanks! Will fix accordingly.

On related note, I'm planning to fix similarly the pinmux driver. In CEC the gpio and pinmux is basically the same thing, and it needs the same address / reserved-ranges list. Any suggestions if sharing this data is possible? And how to group them in DTS? E.g. should the gpio be child of pinmux controller or otherway around?

fabled commented:
> > We need to look at supporting `gpio-reserved-ranges`. See https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt#L150
> 
> Thanks! Will fix accordingly.

Actually, this is rather cumbersome, as it gives ranges and not a bitmask. Perhaps convert it to a bitmask when generating #defines from the DTS? I'll try to get that done in separate commit first.

galak committed:
dts/bindings: Update gpio-controller to clarify ngpios

Provide more docs on ngpios and try and clarify the case if there are
holes in the GPIO slot range how ngpios should work.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts/bindings: Update gpio-controller to add additionl properties

Add gpio-reserved-ranges & gpio-line-names properties.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
> > > We need to look at supporting `gpio-reserved-ranges`. See https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/gpio/gpio.txt#L150
> > 
> > 
> > Thanks! Will fix accordingly.
> 
> Actually, this is rather cumbersome, as it gives ranges and not a bitmask. Perhaps convert it to a bitmask when generating #defines from the DTS? I'll try to get that done in separate commit first.

Agreed we should look at updating the generation script to create a mask.

galak commented:
Added #21277

fabled commented:
> Agreed we should look at updating the generation script to create a mask.
> Added #21277

Thanks. Would you like me to look at doing the mask generation? If yes, should I add a new type such as `bit-range-array` or some other similar name? Or perhaps keep it as `array` type what have in the commit, and special case the `gpio-reserved-ranges` property name or can you suggest another way to detect when the mask generation is needed?

franciscomunoz commented:
All our drivers have been extensively validated and we already manufactured proprietary HW based  on the MEC1501.  We plan to finalize our product  in January and we need our drivers to be stable. This is why I would  suggest that you discuss with @scottwcpg the creation of a new HAL for the cec1702 and use the existing drivers. There are several examples already, for example the imx HAL

fabled commented:
@franciscomunoz 

> All our drivers have been extensively validated and we already manufactured proprietary HW based on the MEC1501. We plan to finalize our product in January and we need our drivers to be stable.

Thanks. Then you probably appreciate my work and review your work. Because it seems your validation is not perfect. See e.g. https://github.com/zephyrproject-rtos/zephyr/pull/20075#issuecomment-563365598. Seems the SPI driver semantics are wrong, and if you have code using that, the application code you are writing is also wrong and will break when the SPI driver is fixed to follow Zephyr API semantics.

And please do not understand me wrong. I appreciate all the work you have put into bringing all this to Zephyr. I just want to help to improve it so it's als suitable to use with CEC.

> This is why I would suggest that you discuss with @scottwcpg the creation of a new HAL for the cec1702 and use the existing drivers. There are several examples already, for example the imx HAL

Yes, I would be happy to do that. I have several questions regarding how the MEC HAL was created. If it's manual work, I'm happy to fix it support also CEC, but I also then have some other questions and comments on how to make it even more suitable.

fabled committed:
dts: parse ngpios and gpio-reserved-ranges to define gpio-pin-mask

This adds logic to construct a valid pin mask define directly to
DTS tree based on the ngpios and gpio-reserved-ranges properties.

This simplifies and allows using gpio-reserved-ranges to properly
configure the valid gpio pins in gpio drivers.

Signed-off-by: Timo Teräs <timo.teras@iki.fi>

fabled commented:
Any idea why the `Kconfig` check is failing?

fabled commented:
Shippable test failed due to known timer issue, looks otherwise ok.

galak reviewed (changes requested):
Can you pull the gpio-line-names & gpio-reserved-ranges changes into their own PR.  

We probably want to enhance edtlib.py with a 'GPIO controller' class or something that comprehends these properties correctly.

I'll say that we probably don't want to add a new `type` for bit-ranges.

fabled commented:
> Can you pull the gpio-line-names & gpio-reserved-ranges changes into their own PR.

The are taken as is from the PR #21277 you have submitted earlier.

> We probably want to enhance edtlib.py with a 'GPIO controller' class or something that comprehends these properties correctly.
> 
> I'll say that we probably don't want to add a new `type` for bit-ranges.

Ok? Could you please elaborate slightly this, I'm happy to write the code unless you have the cycles and willingness.

fabled commented:
Do you mean something like:
```
    def _init_gpio(self):
        # Parses self.gpio_reserved_ranges propery

        node = self._node
        if "gpio-reserved-ranges" not in node.props:
            return

        rprop = node.props["gpio-reserved-ranges"]
        nums = rprop.to_nums()
        mask = 0
        for a in range(0, len(nums), 2):
             mask |= (1 << nums[a] + nums[a + 1]) - 1 ^ (1 << nums[a]) - 1

        prop = Property()
        prop.node = self
        prop.name = rprop.name
        prop.description = None
        prop.val = mask
        prop.type = "int"
        prop.enum_index = None
        self.props[prop.name] = prop
```

And just call that to convert the property?

galak commented:
> Do you mean something like:
> 
> ```python
>     def _init_gpio(self):
>         # Parses self.gpio_reserved_ranges propery
> 
>         node = self._node
>         if "gpio-reserved-ranges" not in node.props:
>             return
> 
>         rprop = node.props["gpio-reserved-ranges"]
>         nums = rprop.to_nums()
>         mask = 0
>         for a in range(0, len(nums), 2):
>              mask |= (1 << nums[a] + nums[a + 1]) - 1 ^ (1 << nums[a]) - 1
> 
>         prop = Property()
>         prop.node = self
>         prop.name = rprop.name
>         prop.description = None
>         prop.val = mask
>         prop.type = "int"
>         prop.enum_index = None
>         self.props[prop.name] = prop
> ```
> 
> And just call that to convert the property?

yes, something like that.

galak commented:
> yes, something like that.

I haven't looked at the code in a bit, but need to see what data / info we'd keep with the 'Controller' Object vs the 'gpio' object.

fabled commented:
New draft proposal based on input. Please review if it's better or needs still adjustments. Thanks.

mnkp commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -329,260 +309,278 @@ static const struct gpio_driver_api gpio_xec_driver_api = {
+		POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		&gpio_xec_driver_api);
+
+#endif /* DT_INST_0_MICROCHIP_XEC_GPIO */
```
In the increasing number of the device drivers we use macros to define driver instances. This update may be a good opportunity to switch to this new style since there is a lot of repeated code. One example could be found in the `drivers/gpio/gpio_nrfx.c` driver.

fabled replied:
Unfortunately this is not possible since we have `#ifdef` conditions for some of the code. It would need to be repeated, or be constructed using templates similar to  `drivers/serial/uart_ns16550_port_x.h` which involves CMake doing the replacements. I thought it was simpler to leave it repeated for the time being. It might be follow up commit when the GIRQ aggregator is implemented properly and no `#ifdef` is needed to do the merging.

fabled replied:
But infact, the interrupt enable should depend on DT `IRQ` field, and DT GIRQ should go directly to GIRQ. So they are two separate things. I'll fixup this. This should also show why the earlier mentioned irq check makes sense.

mnkp reviewed (commented):
I reviewed "gpio: mchp xec: make fully DTS configurable" commit only. LGTM with minor comments.

fabled commented on drivers/gpio/gpio_mchp_xec.c:
```diff
@@ -155,7 +131,9 @@ static int gpio_xec_pin_interrupt_configure(struct device *dev,
-		MCHP_GIRQ_ENCLR(config->girq_id) = BIT(pin);
+		if (config->girq_id) {
+			MCHP_GIRQ_ENCLR(config->girq_id) = BIT(pin);
+		}
```
Technically IRQ and GIRQ are independent. The GIRQ is separate interrupt aggregator, and might be possible that SOCs come out that do not use it, and connect directly to ARM IRQs. However, such SOCs do not exists currently.

Ideally the GIRQ would be separate interrupt controller block, and we'd just route the interrupts in DTS to it, so we'd get rid of all the GIRQ extra variables. Work for this is in #16068.

Until this happens, maybe it's worth keeping the separation and checks here?

mnkp replied:
Sounds good.

galak reviewed (commented):
Sorry, was swamped in trying to get PR #23245.  Can you pull the edtlib.py and dts/bindings/gpio/gpio-controller.yaml changes into their own PR.

I'd like to see about us adding `gpio-line-names` support as well into edtlib.py.

fabled commented:
> Sorry, was swamped in trying to get PR #23245. Can you pull the edtlib.py and dts/bindings/gpio/gpio-controller.yaml changes into their own PR.

Created PR #23660

> I'd like to see about us adding `gpio-line-names` support as well into edtlib.py.

That sounds like a bit of more work, and worth as separate PR. I'd like to finish the reserved-ranges support and the XEC GPIO fixes first. Thanks.
